src/sources/prometheus/mod.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{topology::config::GlobalOptions, Event};
+use futures::{sync::mpsc, Future, Sink, Stream};
+use http::Uri;
+use hyper;
+use hyper_tls::HttpsConnector;
+use serde::{Deserialize, Serialize};
+use std::time::{Duration, Instant};
+use tokio::timer::Interval;
+
+pub mod parser;
+
+#[derive(Deserialize, Serialize, Clone, Debug)]
+struct PrometheusConfig {
+    host: String,
+    #[serde(default = "default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+}
+
+pub fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+#[typetag::serde(name = "prometheus")]
+impl crate::topology::config::SourceConfig for PrometheusConfig {
+    fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        Ok(prometheus(self.clone(), out))
+    }
+
+    fn output_type(&self) -> crate::topology::config::DataType {
+        crate::topology::config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "prometheus"
+    }
+}
+
+fn prometheus(config: PrometheusConfig, out: mpsc::Sender<Event>) -> super::Source {
+    let out = out.sink_map_err(|e| error!("error sending metric: {:?}", e));
+
+    let task = Interval::new(
+        Instant::now(),
+        Duration::from_secs(config.scrape_interval_secs),
+    )
+    .map_err(|e| error!("timer error: {:?}", e))
+    .map(move |_| {
+        let uri = format!("{}/metrics", config.host).parse::<Uri>().unwrap();
+        let request = hyper::Request::get(uri).body(hyper::Body::empty()).unwrap();
+
+        let https = HttpsConnector::new(4).expect("TLS initialization failed");
+        let client = hyper::Client::builder().build(https);
+
+        client
+            .request(request)
+            .and_then(|response| response.into_body().concat2())
+            .map(|body| {
+                let packet = String::from_utf8_lossy(&body);
+                let metrics = parser::parse(&packet)
+                    .map_err(|e| error!("parsing error: {:?}", e))
+                    .unwrap_or_default()
+                    .into_iter()
+                    .map(Event::Metric);
+                futures::stream::iter_ok(metrics)
+            })
+            .flatten_stream()
+            .map_err(|e| error!("http request processing error: {:?}", e))
+    })
+    .flatten()
+    .forward(out)
+    .map(|_| info!("finished sending"));
+
+    Box::new(task)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        sinks::prometheus::PrometheusSinkConfig,
+        test_util::{block_on, next_addr, runtime},
+        topology::{self, config},
+    };
+    use hyper::service::{make_service_fn, service_fn_ok};
+    use hyper::{Body, Response, Server};
+    use pretty_assertions::assert_eq;
+    use std::{thread, time::Duration};
+
+    #[test]
+    fn test_prometheus_routing() {
```

Comment:
I can't quite tell what this is testing. It seems like we set up an example metrics endpoint and then query it, but don't seem to exercise the actual source at all.

Reply:
The idea of the test is to create a [topology](https://github.com/timberio/vector/blob/metrics-prom/src/sources/prometheus/mod.rs#L164) which will scrape metrics from some prometheus-like endpoint, and create a new exposition. Ideally both input and output expositions should be the equal :) (but they are not)

Reply:
Has this been resolved? This seems like a great test to ensure our data model is accurate. We should do similar tests for different combinations of sources/sinks. Do you agree?

Reply:
Yes, this makes sense. I'm not quite sure about the effectiveness and cost of testing all combinations, but I think that testing X -> X workflows is a good idea to begin with. This might give confidence that Vector could be plugged into some existing metric workflow (as transparent proxy) and will not drop any important data.